### DISCUSSION TITLE
How to use reranking models (e.g., jina-reranker-v2) with Hugot?

### DISCUSSION THREAD

**Discussion by Bornholm** [knights-analytics/hugot#75] [Q&A] [answered]
Hi everyone!

First of all, thank you for this wonderful project! I'm currently considering integrating it into one of my pet projects (a simple RAG service) and I have some questions about how to use certain models from Hugging Face.

For example, how could I use a reranking model like jinaai/jina-reranker-v2-base-multilingual? I've run some tests but the results don't seem correct. Could anyone provide some guidance or pointers?

Thanks in advance!

**Comment by riccardopinosio** [comment, marked as answer]
Hi there! Thank you for your interest.

Short answer is this is currently not (yet) supported in that the FeatureExtractionPipeline is designed for (sentence) embedding, e.g. like all-miniLM from sentence transformers, not reranking (i.e. cross encoder) models like the Jina one (which rerank rather than producing embeddings).

Implementing cross encoders for hugot should not be hard however, it amounts to basically converting the logic of the `AutoModelForSequenceClassification` huggingface AutoModel to replicate what they do e.g. here https://huggingface.co/jinaai/jina-reranker-v1-tiny-en in the huggingface transformers approach, creating a new pipelineType like e.g. `CrossEncoderPipeline`.

We had the request to use CrossEncoder models before in the past so it might be something I can take a look at, can't provide timelines though - might have some time this weekend

**Reply by Bornholm** [reply]
Thank you very much for your quick response! I'll put this aside for now and revisit it later. I appreciate your help!

**Reply by riccardopinosio** [reply]
Hi @Bornholm, I added the reranker pipeline on the dev branch of hugot, if you have a chance give it a try and see if it works for your use case.

**Reply by Bornholm** [reply]
@riccardopinosio Thanks a lot; I’ll try to test it, hopefully this week!

**Reply by Hoekz** [reply]
@riccardopinosio I tried this out with the model [Xenova/ms-marco-MiniLM-L-6-v2](https://huggingface.co/Xenova/ms-marco-MiniLM-L-6-v2) and while I could get it to run, the emitted scores were quite different from what I saw running the model using JS and Python. In general, the reranked order was reasonable (top result matched for all the tests I ran) but there were discrepancies in the order and the scores were much lower (0.04 for Golang vs. 0.9999 in JS and Python). Looking at the logits post run, they are all negative during every run. I'm not an expert on this by any means so I have not been able to diagnose where the discrepancies arise.

```go
type RerankRequest struct {
	Query     string   `json:"query" doc:"The query to use for reranking"`
	Documents []string `json:"documents" doc:"The documents to be reranked"`
}

type ScoredDocument struct {
	Document string  `json:"document" doc:"The reranked document"`
	Score    float64 `json:"score" doc:"The reranking score"`
}

type RankedDocuments struct {
	Documents []ScoredDocument `json:"documents" doc:"The list of reranked documents"`
}

func rerank(ctx context.Context, req *RerankRequest) (*RankedDocuments, error) {
	if ctx == nil {
		return nil, errs.New("context is nil")
	}

	if req.Query == "" {
		return nil, errs.New("query is empty")
	}

	if len(req.Documents) == 0 {
		return nil, errs.New("documents is empty")
	}

	session, err := hugot.NewGoSession()
	if err != nil {
		return nil, errs.Wrap(err, "failed to create hugot session")
	}
	defer session.Destroy()

	options := hugot.NewDownloadOptions()
	options.OnnxFilePath = "onnx/model.onnx"
	modelPath, err := hugot.DownloadModel("Xenova/ms-marco-MiniLM-L-6-v2", "./models/", options)
	if err != nil {
		return nil, errs.Wrap(err, "failed to download model")
	}

	config := hugot.CrossEncoderConfig{
		ModelPath:    modelPath,
		OnnxFilename: "model.onnx",
		Name:         "reranker",
	}

	pipeline, err := hugot.NewPipeline(session, config)
	if err != nil {
		return nil, errs.Wrap(err, "failed to create pipeline")
	}

	results, err := pipeline.RunPipeline(req.Query, req.Documents)
	if err != nil {
		return nil, errs.Wrap(err, "failed to run pipeline")
	}

	docs := make([]ScoredDocument, len(results.Results))
	for i, result := range results.Results {
		docs[i] = ScoredDocument{
			Document: result.Document,
			Score:    float64(result.Score),
		}
	}

	return &RankedDocuments{Documents: docs}, nil
}
```

Running it with the example from HuggingFace, you can see the discrepancy:
```sh
curl 'http://[::1]:8644/rerank' \
  --request POST \
  --header 'Content-Type: application/json' \
  --data '{
  "documents": [
    "Berlin had a population of 3,520,031 registered inhabitants in an area of 891.82 square kilometers.",
    "Berlin is well known for its museums."
  ],
  "query": "How many people live in Berlin?"
}'
# {
#   "documents": [
#    {
#       "document": "Berlin had a population of 3,520,031 registered inhabitants in an area of 891.82 square kilometers.",
#       "score": 0.004407973960042
#     },
#     {
#       "document": "Berlin is well known for its museums.",
#       "score": 0.00018127336807083338
#     }
#   ]
# }
# Observed logits: [ -5.4199224, -8.615323 ]

# Logits on HuggingFace
# quantized:   [ 8.663132667541504, -11.245542526245117 ]
# unquantized: [ 8.845855712890625, -11.245561599731445 ]
# scores: [ 0.999856043542086, 0.000013064986859073432 ]
```

**Reply by riccardopinosio** [reply]
Hi, 

Ok interesting. I am testing it against the jina model and the ranking there is correct, but it might be good to look more closely at this example. In principle the logits should match as they use the same onnx model.

Will investigate and keep you posted.

R.

**Reply by Hoekz** [reply]
I saw the current implementation became part of a release...were you able to verify whether or not there's a problem with the the logits / scores / expected order coming back?

**Reply by riccardopinosio** [reply]
hallo, not yet for this specific model (the jina models seem to return the expected orders there's a cicd test for the tiny reranker but i also get the expected ranking for the v2 base). Will check with this ms-marco. One question, you are using the ORT backend not GOMLX yes?

**Reply by riccardopinosio** [reply]
@Hoekz I believe I found the mismatch issue, as is usually the case it was in the tokenization which assumed BERT tokenization and added extra tokens.

Can you try with this branch: https://github.com/knights-analytics/hugot/tree/cross-encoder-fix

I get matching logits/ranking across both the Jina's and the Marco model.

**Reply by Hoekz** [reply]
I'm still seeing discrepancies on my side...for a JS example:

```js
import { env, AutoTokenizer, AutoModelForSequenceClassification } from '@xenova/transformers'
import os from 'os'

env.cacheDir = os.tmpdir()

const sigmoid = x => 1 / (1 + Math.exp(-x))

export async function rerank(query, documents) {
  if (!documents.length) {
    return []
  }

  const model = await AutoModelForSequenceClassification.from_pretrained('Xenova/ms-marco-MiniLM-L-6-v2')
  const tokenizer = await AutoTokenizer.from_pretrained('Xenova/ms-marco-MiniLM-L-6-v2')

  const encoding = await tokenizer(documents.map(() => query), {
    text_pair: documents.map(({ text }) => text),
    padding: true,
    truncation: true,
  })

  const embeddings = await model(encoding)
  const scores = embeddings.logits.data.map(sigmoid)

  const toSort = documents.map((doc, i) => ({ doc, score: scores[i] }))

  return toSort.sort((a, b) => b.score - a.score).map(({ doc, score }) => ({ ...doc, score }))
}

await rerank('What is the capital of France?', [
  { text: 'The capital of France is Paris.' },
  { text: 'The capital of Germany is Berlin.' },
])
// [
//   {
//     text: 'The capital of France is Paris.',
//     score: 0.9998955130577087
//   },
//   {
//     text: 'The capital of Germany is Berlin.',
//     score: 0.10640902072191238
//   }
// ]
```

Whereas when hitting my Go server:

```sh
curl 'http://[::1]:8644/rerank' \
  --request POST \
  --header 'Content-Type: application/json' \
  --data '{
  "documents": [
    "The capital of France is Paris.",
    "The capital of Germany is Berlin."
  ],
  "query": "What is the capital of France?"
}'

# {
#   "$schema": "https://[::1]:8644/schemas/RankedDocuments.json",
#   "documents": [
#    {
#       "document": "The capital of France is Paris.",
#       "score": 0.1041819229722023
#     },
#     {
#       "document": "The capital of Germany is Berlin.",
#       "score": 0.013263199478387833
#     }
#   ]
# }
```

The scores/logits are different this time around though, running the same query against `0.5.3` I get back:

```json
{
  "$schema": "https://[::1]:8644/schemas/RankedDocuments.json",
  "documents": [
    {
      "document": "The capital of France is Paris.",
      "score": 0.029932620003819466
    },
    {
      "document": "The capital of Germany is Berlin.",
      "score": 0.0033979716245085
    }
  ]
}
```

**Reply by riccardopinosio** [reply]
@Hoekz allright, I missed another thing: BERT type models use a different pattern for the token type IDs, i.e. 0...0,1....1 for the two sequences.
Can you try that branch with the new commit again? I have an example in hugot_test.go for the crossencoder pipeline that does:

query := "Organic skincare products for sensitive skin"
documents := []string{
		"Eco-friendly kitchenware for modern homes",
		"Biodegradable cleaning supplies for eco-conscious consumers",
		"Organic cotton baby clothes for sensitive skin",
		"Natural organic skincare range for sensitive skin",
		"Tech gadgets for smart homes: 2024 edition",
		"Sustainable gardening tools and compost solutions",
		"Sensitive skin-friendly facial cleansers and toners",
		"Organic food wraps and storage solutions",
		"All-natural pet food for dogs with allergies",
		"Yoga mats made from recycled materials",
	}

and I get from go:

expectedBert := []Expected{
		{Document: "Natural organic skincare range for sensitive skin", Score: 0.9997},
		{Document: "Organic cotton baby clothes for sensitive skin", Score: 0.9991},
		{Document: "Sensitive skin-friendly facial cleansers and toners", Score: 0.7979},
		{Document: "Organic food wraps and storage solutions", Score: 0.0046},
		{Document: "Biodegradable cleaning supplies for eco-conscious consumers", Score: 0.0000},
		{Document: "Sustainable gardening tools and compost solutions", Score: 0.0000},
		{Document: "Yoga mats made from recycled materials", Score: 0.0000},
		{Document: "Eco-friendly kitchenware for modern homes", Score: 0.0000},
		{Document: "All-natural pet food for dogs with allergies", Score: 0.0000},
		{Document: "Tech gadgets for smart homes: 2024 edition", Score: 0.0000},
	}

which is consistent with python:

Scores (sigmoid):
0.9997  |  Natural organic skincare range for sensitive skin
0.9991  |  Organic cotton baby clothes for sensitive skin
0.7979  |  Sensitive skin-friendly facial cleansers and toners
0.0046  |  Organic food wraps and storage solutions
0.0000  |  Biodegradable cleaning supplies for eco-conscious consumers
0.0000  |  Sustainable gardening tools and compost solutions
0.0000  |  Yoga mats made from recycled materials
0.0000  |  Eco-friendly kitchenware for modern homes
0.0000  |  All-natural pet food for dogs with allergies
0.0000  |  Tech gadgets for smart homes: 2024 edition

**Reply by Hoekz** [reply]
That seems to have worked! Thanks so much for working on this.

**Reply by riccardopinosio** [reply]
great, will make a patch release for it
